 Samuel Thibault, le Sun 04 May 2008 00:50:27 +0100, a écrit :
 > Willy Tarreau, le Wed 30 Apr 2008 21:49:20 +0200, a écrit :
 > > On Wed, Apr 30, 2008 at 01:08:51AM +0100, Samuel Thibault wrote:
 > > > Willy Tarreau wrote:
 > > > > 3) if I enter Alt-196, I get a "Ä". Flushing the buffer shows that od
 > > > > got two bytes: c3 84.
 > > > 
 > > > Confirmed.
 > > > 
 > > > Try init=/bin/stty -a, that will show
 > > > 
 > > > -iutf8
 > > > 
 > > > So there is little wonder that canonical mode does not work as expected.
 > > > 
 > > > Try init=/bin/sh, from that shell run stty iutf8. Then things will work
 > > > fine.  The fix is thus just to make the VT's tty initial iutf8 setup
 > > > follow vt.default_utf8.
 > > 
 > > Will try that on a more recent install. Mine's stty does not support
 > > this option. Your analysis makes quite a lot of sense, and such a fix
 > > would wipe part of my annoyances/anger with this recent change.
 > 
 > Can you give the patch below a try?
 > Dynamic per-VT utf-8 switch should also work, provided that you reopen
 > the VT (i.e. log out).

 Willy Tarreau, le Sun 04 May 2008 10:55:14 +0200, a écrit :
 > I confirm that your patch works perfectly for me. Now backspace correctly
 > removes multi-byte characters. My bash is still fooled though but as Alan
 > explained it, it's readline which has to be upgraded now.

I guess this is suitable for the stable trees of 2.6.24 and 2.6.25
(where UTF-8 is by default now).


Set IUTF8 as appropriate on VT tty open.

Signed-off-by: Samuel Thibault <samuel.thibault@ens-lyon.org>

--- linux/drivers/char/vt.c.orig	2008-05-04 00:37:50.000000000 +0100
+++ linux/drivers/char/vt.c	2008-05-04 00:47:39.000000000 +0100
@@ -2723,6 +2723,10 @@ static int con_open(struct tty_struct *t
 				tty->winsize.ws_row = vc_cons[currcons].d->vc_rows;
 				tty->winsize.ws_col = vc_cons[currcons].d->vc_cols;
 			}
+			if (vc->vc_utf)
+				tty->termios->c_iflag |= IUTF8;
+			else
+				tty->termios->c_iflag &= ~IUTF8;
 			release_console_sem();
 			vcs_make_sysfs(tty);
 			return ret;
@@ -2899,6 +2903,8 @@ int __init vty_init(void)
 	console_driver->minor_start = 1;
 	console_driver->type = TTY_DRIVER_TYPE_CONSOLE;
 	console_driver->init_termios = tty_std_termios;
+	if (default_utf8)
+		console_driver->init_termios.c_iflag |= IUTF8;
 	console_driver->flags = TTY_DRIVER_REAL_RAW | TTY_DRIVER_RESET_TERMIOS;
 	tty_set_operations(console_driver, &con_ops);
 	if (tty_register_driver(console_driver))

